egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decode.py

```diff
@@ -0,0 +1,607 @@
+    )
+    params.res_dir = params.exp_dir / params.decoding_method
+
+    params.suffix = f"epoch-{params.epoch}-avg-{params.avg}"
+    if "fast_beam_search" in params.decoding_method:
+        params.suffix += f"-beam-{params.beam}"
+        params.suffix += f"-max-contexts-{params.max_contexts}"
+        params.suffix += f"-max-states-{params.max_states}"
+    elif "beam_search" in params.decoding_method:
+        params.suffix += f"-beam-{params.beam_size}"
+    else:
+        params.suffix += f"-context-{params.context_size}"
+        params.suffix += f"-max-sym-per-frame-{params.max_sym_per_frame}"
+
+    setup_logger(f"{params.res_dir}/log-decode-{params.suffix}")
+    logging.info("Decoding started")
+
+    device = torch.device("cpu")
+    if torch.cuda.is_available():
+        device = torch.device("cuda", 0)
+
+    logging.info(f"Device: {device}")
+
+    lexicon = Lexicon(params.lang_dir)
+    params.blank_id = lexicon.token_table["<blk>"]
+    params.vocab_size = max(lexicon.tokens) + 1
+
+    logging.info(params)
+
+    logging.info("About to create model")
+    model = get_transducer_model(params)
+
+    if params.avg_last_n > 0:
+        filenames = find_checkpoints(params.exp_dir)[: params.avg_last_n]
+        logging.info(f"averaging {filenames}")
+        model.to(device)
+        model.load_state_dict(average_checkpoints(filenames, device=device))
+    elif params.avg == 1:
+        load_checkpoint(f"{params.exp_dir}/epoch-{params.epoch}.pt", model)
+    elif params.batch is not None:
+        filenames = f"{params.exp_dir}/checkpoint-{params.batch}.pt"
+        logging.info(f"averaging {filenames}")
+        model.to(device)
+        model.load_state_dict(average_checkpoints([filenames], device=device))
+    else:
+        start = params.epoch - params.avg + 1
+        filenames = []
+        for i in range(start, params.epoch + 1):
+            if start >= 0:
+                filenames.append(f"{params.exp_dir}/epoch-{i}.pt")
+        logging.info(f"averaging {filenames}")
+        model.to(device)
+        model.load_state_dict(average_checkpoints(filenames, device=device))
+
+        average = average_checkpoints(filenames, device=device)
+        checkpoint = {"model": average}
+        torch.save(
+            checkpoint,
+            "pruned_transducer_stateless2/pretrained_average_11_to_29.pt",
+        )
+
+    model.to(device)
+    model.eval()
+    model.device = device
+
+    if params.decoding_method == "fast_beam_search":
+        decoding_graph = k2.trivial_graph(params.vocab_size - 1, device=device)
+    else:
+        decoding_graph = None
+
+    num_param = sum([p.numel() for p in model.parameters()])
+    logging.info(f"Number of model parameters: {num_param}")
+
+    # Note: Please use "pip install webdataset==0.1.103"
+    # for installing the webdataset.
+    import glob
+    import os
+
+    from lhotse import CutSet
+    from lhotse.dataset.webdataset import export_to_webdataset
```

Comment:
Oh, I use webdataset here.

Reply:
Hmm interesting. Were you running multi-GPU decoding? In that case you might find that updating to latest Lhotse will help, I recently fixed an issue that only appears with multi-GPU and multi-dataloader-worker scripts.

Reply:
> Were you running multi-GPU decoding?

I don't think so. All the decoding code in icefall uses a single GPU.

Is it expected that if `--num-workers=2` in this case, the data is doubled?

Reply:
No, that shouldn't happen since he used the shards option in exporting to webdataset. If you can confirm it's really because of webdataset, and updating to latest Lhotse (and webdataset) doesn't help, we'll debug it.

Reply:
OK.

Reply:
I use the latest Lhotse and webdataset to decode again. When I set num-workers as 2 and a single GPU, the data is still doubled.

Reply:
Also, it works fine when number of workers is 1 or 0, I think.

Reply:
OK I think what happened. Replace `BucketingSampler` with `DynamicBucketingSampler` and it should resolve the issues. With `BucketingSampler`, the cuts are eagerly read into memory before they go into dataloader's sub-processes, so webdataset doesn't know to de-duplicate them. I believe you are seeing a warning about BucketingSampler not being a good fit for lazy cutsets; now I'm thinking I should make it raise an exception saying to migrate to DynamicBucketingSampler in these cases to avoid such issues.

Reply:
Ah, yes. `luomingshuang/anaconda3/envs/k2-python/lib/python3.8/site-packages/lhotse-1.2.0.dev0+git.de75634.clean-py3.8.egg/lhotse/dataset/sampling/bucketing.py:102: UserWarning: Lazy CutSet detected in BucketingSampler: we will read it into memory anyway. Please use lhotse.dataset.DynamicBucketingSampler instead.
`

Reply:
See: https://github.com/lhotse-speech/lhotse/pull/710

Reply:
But, isn't DynamicBucketSampler slow down when loading data?

Reply:
I think you are referring to issue https://github.com/lhotse-speech/lhotse/issues/678 -- here it's a different use-case, because the sampler itself is moved to the dataloading worker process via `IterableDatasetWrapper`, so any overhead introduced by it doesn't affect the main training loop.

Reply:
OK, when using DynamicbucketSampler for decoding, the number of cuts is normal (not doubled) even if the num-workers is 2.